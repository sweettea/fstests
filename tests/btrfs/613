#! /bin/bash
# SPDX-License-Identifier: GPL-2.0
# Copyright (c) 2023 Meta Platforms, Inc.  All Rights Reserved.
#
# FS QA Test 613
#
# Check if reflinking one encrypted file on btrfs succeeds.
#
. ./common/preamble
_begin_fstest auto encrypt

# Import common functions.
. ./common/encrypt
. ./common/filter
. ./common/reflink

# real QA test starts here

# Modify as appropriate.
_supported_fs btrfs

_require_test
_require_scratch
_require_cp_reflink
_require_scratch_encryption -v 2
_require_command "$KEYCTL_PROG" keyctl

_scratch_mkfs_encrypted &>> $seqres.full
_scratch_mount

dir=$SCRATCH_MNT/dir
mkdir $dir
_set_encpolicy $dir $TEST_KEY_IDENTIFIER
_add_enckey $SCRATCH_MNT "$TEST_RAW_KEY"
echo "Creating and reflinking a file"
$XFS_IO_PROG -t -f -c "pwrite 0 33k" $dir/test > /dev/null
cp --reflink=always $dir/test $dir/test2

echo "Can't reflink encrypted and unencrypted"
cp --reflink=always $dir/test $SCRATCH_MNT/fail |& _filter_scratch

echo "Diffing the file and its copy"
diff $dir/test $dir/test2

echo "Verifying the files are reflinked"
_verify_reflink $dir/test $dir/test2

echo "Diffing the files after remount"
_scratch_cycle_mount
_add_enckey $SCRATCH_MNT "$TEST_RAW_KEY"
diff $dir/test $dir/test2

echo "Diffing the files after key remove"
_rm_enckey $SCRATCH_MNT $TEST_KEY_IDENTIFIER
diff $dir/test $dir/test2 |& _filter_scratch

# success, all done
status=0
exit
