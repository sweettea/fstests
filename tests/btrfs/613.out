QA output created by 613
Added encryption key with identifier 69b2f6edeee720cce0577937eb8a6751
Creating and reflinking a file
Can't reflink encrypted and unencrypted
cp: failed to clone 'SCRATCH_MNT/fail' from 'SCRATCH_MNT/dir/test': Invalid argument
Diffing the file and its copy
Verifying the files are reflinked
Diffing the files after remount
Added encryption key with identifier 69b2f6edeee720cce0577937eb8a6751
Diffing the files after key remove
Removed encryption key with identifier 69b2f6edeee720cce0577937eb8a6751
diff: SCRATCH_MNT/dir/test: No such file or directory
diff: SCRATCH_MNT/dir/test2: No such file or directory
